src/QsCompiler/Core/Transformations/ExpressionTransformation.fs

```diff
@@ -316,10 +316,18 @@ type ExpressionKindTransformationBase internal (options: TransformationOptions,
     abstract OnLambda : lambda: TypedExpression Lambda -> ExpressionKind
 
     default this.OnLambda lambda =
+        let rec onSymbol s =
```

Comment:
I don't really see a good place to put a unified OnSymbolTuple handler - it can occur in both statements and expressions but doesn't belong to either. Where would you prefer to move it?

Reply:
How about putting it in the NamespaceTransformation? This seems appropriate given that other symbols (declarations) also occur there, and at some point we might have something like global (constant) variables.

Reply:
In terms of implementation, I believe this will require making the expression, statement and namespace transformations mutually recursive (i.e., ExpressionTransformation.fs, StatementTransformation.fs, and NamespaceTransformation.fs need to be combined into one file).

In terms of API design, I'm not sure I agree that symbol tuples have an association with namespaces. IMO, they have no particular association with types, expressions, statements, or namespaces, and I think any choice of where we put it is arbitrary. I would still recommend a single unified transformation class as a simpler design alternative to the current transformations.

Reply:
I think this would need to make the expression, statement and namespace transformations mutually recursive (i.e., ExpressionTransformation.fs, StatementTransformation.fs, and NamespaceTransformation.fs need to be combined into one file).

Reply:
I'm not sure I agree with putting it in NamespaceTransformation though... It's true that declarations occur in namespaces, but they can also occur in statements and expressions, so favoring NamespaceTransformation feels arbitrary to me. Plus, SymbolTuple specifically is an untyped collection of symbols, which is one type of declaration that we potentially may never allow at the namespace level.

I think this is an example of how having fewer transformation classes would make it easier to organize things - for example, if Q# was expression-based, we'd have one transformation for both expressions and (former) statements, so OnSymbolTuple can go there. Or if we had one unified transformation class. In my experience, syntax trees often have little utility nodes like SymbolTuple that don't fit neatly into the "namespace, statement, expression, type" divisions.

Reply:
I created a PR with a clean up to have some consistency in the transformations: https://github.com/microsoft/qsharp-compiler/pull/1290

Reply:
I'll approve this PR and expect to have #1290 updated later today to do a slightly more extensive revision of the transformations.